README.md
#  Expert Lab: live chat application

this is a application meant to be used as a live chat application for a website. 
this would be acomplished by using web sockets to connect to a server and send and recieve messages.
the user could get a random username and send messages to each other, as soon as the user disconnects all his messages will be deleted.

## Table of content
- [Expert Lab: live chat application](#web-ii-front-end-brusselaers-niko)
  - [Table of content](#table-of-content)
  - [Installation](#installation)
    - [Requirements](#requirements)
    - [Setup](#setup)
  - [links](#links)
    - [used references:](#used-references)


## Installation
### Requirements
to install and run this application you will need the following:
  - node: https://nodejs.org/en/

### Setup
to install and run this application you will need the following:
  - clone this repository
  - run `npm install` to install all the dependencies
  - run `npm start` to start the application

## links
    - trello: https://trello.com/b/tWRLhGkV/sprint-1-live-chat-application   

### used references:
    - chatGPT: https://chat.openai.com/ 
        (used to generate the content and json files)
    - socket.io: https://socket.io/
        (used to create the websockets)
    - video tutorial: https://trello.com/b/tWRLhGkV/sprint-1-live-chat-application
        (example way to build live chat app with socket.io)
    - video tutorial socket io postman: https://trello.com/b/tWRLhGkV/sprint-1-live-chat-application
        (tutorial on how to use socket.io with postman)
    - tailwindcss: https://tailwindcss.com/
        (used to style the application)

changelog.md
# changelog

## [0.0.1] - 2022-10-10

### Added

    created initial frontend structure of live chat application
    -frontend
        -sideNavComponent
        -chatComponent
        -messageComponent
        -messageInputComponent
    
## [0.0.2] - 2022-11-10

### Added

    created backend server with socket io and added sockets:
    - login : used to login with username
    - logout : used to logout with username

    started on building socket:
    - message

## [0.0.3] - 2022-20-10
    
    ### changed

    - message socket: made it possible to make chatrooms and send messages to specific chatrooms
    - login socket: made it so that socketId is saved with username in database

    ### Added

    - getmessages socket : way to let user get updated messages and chatrooms from database
    - disconnect socket : way to let user disconnect from chatroom and where user and user data is removed from database

    ### removed

    - logout socket 

## [0.0.4] - 2022-21-10

    ### changed

    - frontend: updated style of frontend live chat application

images/backend/src/index.js
import express from 'express';
import http from 'http';
import { Server } from 'socket.io';

const app = express();
const server = http.createServer(app);
const Port = process.env.PORT || 3000;
const io = new Server(server)

const Users = [];

server.listen(3000, () => {
    console.log(`listening on http://localhost:${Port}`);
});

io.on('connection', (socket) => {
    
    // socket to make new user
    socket.on('login', (data) => {

        // check if user already exists
        if (!Users.includes(data.newUser)) {
            console.log("new user added");
            // add new user to users array
            Users.push(data.newUser);
            let response = {
                data: "new user added",
                statuscode: 200
            }
            // send response to client
            socket.emit('login', response);
        } else {
            console.log("user already exists");
            let response = {
                data: "user already exists",
                statuscode: 401
            }
            // send error response to client
            socket.emit('login', response);

        }

    })

    // socket to remove user
    socket.on('logout', (data) => {
        Users.splice(Users.indexOf(data.user), 1);
        console.log("user removed");
        // send response to client
        socket.emit('logout', "user removed");
    })

    



});





images/backend/src/index.ts
import express from 'express';
import http from 'http';
import { Server, Socket } from 'socket.io';
import { IUser } from '../shared/models/user';
import { IChatroom } from '../shared/models/chatroom';
import { IMessage } from '../shared/models/message';

const app = express();
const server = http.createServer(app);
const Port = process.env.PORT || 3000;
const io = new Server(server)

const users:IUser[] = [];
const chatrooms:IChatroom[] = [];
let newChatroomID = 0;
let newMessageID = 0;


server.listen(3000, () => {
    console.log(`listening on http://localhost:${Port}`);
});

io.on('connection', (socket) => {    
    // socket to make new user

    socket.on('login', (data) => {
        // check if user already exists
        if (users.find(user => user.username === data.newUser) === undefined 
            && users.find(user => user.id === socket.id) === undefined) 
        {
            console.log("new user added");            
            // add new user to users array
            users.push(
                {
                    id: socket.id,
                    username: data.newUser
                }
            );
            let response = {
                data: "new user added",
                userId: socket.id,
                statuscode: 200
            }
            // send response to client
            socket.emit('login', response);

        } else {
            console.log("user already exists");
            let response = {
                data: "user already exists",
                statuscode: 401
            }
            // send error response to client
            socket.emit('login', response);

        }
        

    })

    

    // socket to send message to other users
    socket.on('message', (data) => {
        try {
            
        // get chatroom if one exists
        let chatroom = chatrooms.find(chatroom => chatroom.id === data.chatroomID);
        
        if (!chatroom) {
            // create new chatroom if it doesn't exist
            chatroom = createNewChatroom(data.chatroomName, [data.sender, ...data.receiver]);
        }
        // add message to chatroom
        chatroom.messages.push(
            {
                id: newMessageID,
                sender: data.sender,
                message: data.message
            }
        )
        newMessageID++;

        
        // send response to client
        socket.broadcast.to(findIdByName(chatroom.users)).emit('message', chatroom);
        

        } catch (error) {
            //catch error
            console.log(error);
        }        
    });

    // socket to remove user when discconected
    socket.on('disconnect', () => {
        // remove user from users array
        let user = users.find(user => user.id === socket.id);
        users.splice(users.findIndex(user => user.id === socket.id), 1);
        //find all chatrooms where disconnected user is a member
        let chatroomsToRemove = chatrooms.filter(chatroom => chatroom.users.includes(user?.username as string));
        // remove user from chatrooms
        chatroomsToRemove.forEach(chatroom => {
            chatroom.users.splice(chatroom.users.findIndex(username => username === user?.username), 1);
            //if chatroom only has 1 user left, remove chatroom
            if (chatroom.users.length === 1) {
                chatrooms.splice(chatrooms.findIndex(chatroom => chatroom.id === chatroom.id), 1);
            }
            //broadcoast to other user of chatroom that user has disconnected
            socket.broadcast.to(findIdByName(chatroom.users)).emit('disconnected', user?.username);
        });

        
    })

    // socket to get all messages from current user
    socket.on('getMessages', (data) => {
        // find all chatrooms where current user is a member
        let userChatrooms = chatrooms.filter(chatroom => chatroom.users.includes(data.username));
        // return all chatrooms to user
        socket.emit('getMessages', userChatrooms);
    })

});



function createNewChatroom(name:string, users:string[]):IChatroom{
    // create new chatroom and save to chatrooms array
    let newChatroom = {
        id: newChatroomID,
        name: name,
        users: users,
        messages: []
    }
    chatrooms.push(newChatroom);
    newChatroomID++;

    return newChatroom;
}

function findIdByName(usernames: string[]): string[] {
    // find id of user by username
    let ids = usernames.map((username: string) => (users.find(user => user.username === username)?.id ).toString());
    return ids;
}


